### PR TITLE
[Workflows] Migrate to Symfony Workflow Component

### LINKED ISSUE #3166: [Workflows] Migrate to Symfony Workflow Component
- Workflow based on Symfony Workflow Engine - either as workflow or as state machine 
- Multiple workflows per element (object, document, asset)
- Definition of workflows in yml config file and validation with Symfony config tree
- Configuration of support strategy that defines which elements are allowed for that workflow
- Configuration of marking store that is responsible for storing workflow state of elements - either extra table or attribute in Pimcore object
- Configuration of places (former states & statuses) with label, color, permissions
- Configuration of transitions with guards (is transition allowed for certain object), from, to and additional options like label, icon class, additional notes
- Graphical representation of workflow (read only) and place of element

### PR DESCRIPTION
**Pull request by markus-moser**
This pull request will resolve #3166

### CODE REVIEW

File: doc/Development_Documentation/07_Workflow_Management/README.md [hunk 1]
```diff
@@ -65,19 +65,39 @@ array of user(s) or role(s) that you would like to be notified in options sectio
...
 
+Multiple notification settings with conditions allow sophisticated notifications to be configured for each transition. 
 To customize the e-mail template, following options are available: 
-- Overwrite the template `@PimcoreCore/Workflow/NotificationEmail/notificationEmail.html.twig`. Default parameters 
-available in the template are `subjectType`, `subject`, `action`, `deeplink`, `note_description`, `translator`, `lang`. 
-- If additional parameters are required, overwrite the service `Pimcore\Workflow\NotificationEmail\NotificationEmailService`.
+- Overwrite the template `@PimcoreCore/Workflow/NotificationEmail/notificationEmail.html.twig` or configure your own 
+  template path in settings. Default parameters available in the template are `subjectType`, `subject`, `action`, `workflow`, 
+  `workflowName`, `deeplink`, `note_description`, `translator`, `lang`. If additional parameters are required, overwrite 
+  the service `Pimcore\Workflow\NotificationEmail\NotificationEmailService`.
```
Comment (dpfaffenbauer) [inline doc/Development_Documentation/07_Workflow_Management/README.md]: or use Twig Extensions?

File: lib/Workflow/NotificationEmail/NotificationEmailService.php [hunk 2]
```diff
@@ -17,14 +17,18 @@
...
 use Symfony\Component\Templating\EngineInterface;
 use Symfony\Component\Routing\RouterInterface;
 use Symfony\Component\Translation\TranslatorInterface;
 use Pimcore\Model\User;
 use Pimcore\Model\Element;
+use Symfony\Component\Workflow\Workflow;
 
 class NotificationEmailService
 {
+    const MAIL_PATH_LANGUAGE_PLACEHOLDER = '$$lang$$';
```
Comment (dpfaffenbauer) [inline lib/Workflow/NotificationEmail/NotificationEmailService.php]: why not %_locale% like symfony does for routes?

File: doc/Development_Documentation/07_Workflow_Management/README.md (hunk 1 above)
Comment (fashxp) [reply doc/Development_Documentation/07_Workflow_Management/README.md]: sure, but if the data is also not available to twig extensions, you still have the possibility to extend the service.

File: bundles/AdminBundle/Controller/Admin/WorkflowController.php [hunk 3]
```diff
@@ -14,21 +14,28 @@
 
 namespace Pimcore\Bundle\AdminBundle\Controller\Admin;
 
+use http\Exception\InvalidArgumentException;
```
Comment (dpfaffenbauer) [inline bundles/AdminBundle/Controller/Admin/WorkflowController.php]: really? not \InvalidArgumentException?

Comment (markus-moser) [reply bundles/AdminBundle/Controller/Admin/WorkflowController.php]: you're right - the import is useless. I dropped it, thx.

File: bundles/CoreBundle/Command/WorkflowDumpCommand.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        $container = $this->getApplication()->getKernel()->getContainer();
+        $serviceId = $input->getArgument('name');
+        if ($container->has('workflow.'.$serviceId)) {
+            $workflow = $container->get('workflow.'.$serviceId);
+            $dumper = $this->getContainer()->get(GraphvizDumper::class);
+        } elseif ($container->has('state_machine.'.$serviceId)) {
+            $workflow = $container->get('state_machine.'.$serviceId);
+            $dumper = $this->getContainer()->get(StateMachineGraphvizDumper::class);
+        } else {
+            throw new InvalidArgumentException(sprintf('No service found for "workflow.%1$s" nor "state_machine.%1$s".', $serviceId));
```
Comment (dpfaffenbauer) [inline bundles/CoreBundle/Command/WorkflowDumpCommand.php]: `\ InvalidArgumentException`

Comment (dpfaffenbauer) [reply bundles/CoreBundle/Command/WorkflowDumpCommand.php]: ah, forgett it, wrong file :D

File: bundles/CoreBundle/Resources/migrations/Version20180904201947.php [hunk 5]
```diff
@@ -17,6 +18,14 @@ public function up(Schema $schema)
...
         $table->setPrimaryKey(['cid', 'ctype', 'workflowId', 'workflow']);
+
+
+        $db = Db::get();
+        try {
+            $db->query("INSERT INTO users_permission_definitions VALUES('workflow_details');");
+        } catch (\Exception $e) {
+        }
+
     }
```
Comment (dpfaffenbauer) [inline bundles/CoreBundle/Resources/migrations/Version20180904201947.php]: `$this->addSql('INSERT INTO users_permission_definitions VALUES('workflow_details');')`

**Pull request merged**